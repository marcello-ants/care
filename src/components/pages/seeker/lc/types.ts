import { GeneralProviderProfile } from '@/types/seeker';

export interface ProviderProfile extends GeneralProviderProfile {
  numberOfReviews: number;
  qualities?: ProviderQualities;
  services?: ProviderServices;
  firstName: string;
}

export interface ProviderQualities {
  certifiedNursingAssistant?: boolean;
  certifiedRegisterNurse?: boolean;
  certifiedTeacher?: boolean;
  childDevelopmentAssociate?: boolean;
  comfortableWithPets?: boolean;
  cprTrained?: boolean;
  crn?: boolean;
  doesNotSmoke?: boolean;
  doula?: boolean;
  earlyChildDevelopmentCoursework?: boolean;
  earlyChildhoodEducation?: boolean;
  firstAidTraining?: boolean;
  nafccCertified?: boolean;
  ownTransportation?: boolean;
  specialNeedsCare?: boolean;
  trustlineCertifiedCalifornia?: boolean;
}

export interface ProviderServices {
  carpooling?: boolean;
  craftAssistance?: boolean;
  errands?: boolean;
  groceryShopping?: boolean;
  laundryAssistance?: boolean;
  lightHousekeeping?: boolean;
  mealPreparation?: boolean;
  swimmingSupervision?: boolean;
  travel?: boolean;
}
