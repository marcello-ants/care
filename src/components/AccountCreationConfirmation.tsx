import { ReactElement } from 'react';
import { Icon64BrandActionSuccess } from '@care/react-icons';
import { Grid, Typography } from '@material-ui/core';
import { StatelessSelector, Pill } from '@care/react-component-lib';
import { makeStyles } from '@material-ui/core/styles';
import Header from '@/components/Header';
import { SelectorType } from '@care/react-component-lib/dist/components/Selector/StatelessSelector';

const useStyles = makeStyles((theme) => ({
  selector: {
    '& .MuiListItem-root': {
      marginBottom: 0,
    },
  },
  icon: {
    position: 'relative',
    left: '-9px',
  },
  bodyText: {
    fontWeight: 'normal',
    lineHeight: '24px',
    margin: theme.spacing(2, 0, 4),
  },
  typeContainer: {
    margin: theme.spacing(1, 0, 1),
  },

  buttonContainer: {
    padding: theme.spacing(3.5, 3, 0),
  },
  header: {
    marginBottom: ({ subheader }: { subheader: string | undefined }) =>
      subheader ? 0 : theme.spacing(1),
  },
}));

type Options = {
  label: string;
  value: string;
};

interface IAccountCreationConfirmation {
  header: string;
  description?: string;
  subheader?: string;
  options: Options[];
  onChange?: SelectorType['onChange'];
  cta?: ReactElement;
  value?: string;
  showIcon?: boolean;
}

function AccountCreationConfirmation({
  header,
  description,
  subheader,
  options,
  value,
  onChange,
  cta,
  showIcon,
}: IAccountCreationConfirmation) {
  const classes = useStyles({ subheader });

  return (
    <Grid container>
      {showIcon && (
        <Grid item xs={12}>
          {/*
          The svg has a litle padding that makes the icon smaller
          60px makes the icon size to 49.33 the correct size is 60px approx
          74px makes the icon size 60.88
        */}
          <Icon64BrandActionSuccess size="74px" className={classes.icon} />
        </Grid>
      )}

      <Grid item xs={12} className={classes.header}>
        <Header>{header}</Header>
      </Grid>
      {description && (
        <Grid item xs={12}>
          <Typography variant="body2" className={classes.bodyText}>
            {description}
          </Typography>
        </Grid>
      )}

      {subheader && (
        <Grid item xs={12}>
          <Typography variant="h4">{subheader}</Typography>
        </Grid>
      )}

      <Grid item xs={12} className={classes.typeContainer}>
        <StatelessSelector onChange={onChange} single value={value} className={classes.selector}>
          {options.map((option) => (
            <Pill key={option.label} {...option} size="md" />
          ))}
        </StatelessSelector>
        {cta && <div className={classes.buttonContainer}>{cta}</div>}
      </Grid>
    </Grid>
  );
}

AccountCreationConfirmation.defaultProps = {
  value: '',
  onChange: () => {},
  cta: null,
  showIcon: true,
  description: '',
  subheader: '',
};

export default AccountCreationConfirmation;
