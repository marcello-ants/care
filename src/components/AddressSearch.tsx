import { useEffect, useState } from 'react';
import { theme } from '@care/material-ui-theme';
import { Box, Button, makeStyles } from '@material-ui/core';
import { GoogleMaps, GoogleMapsActions, useGoogleMaps } from '@care/seeker-react-components';
import { Icon24UtilitySearch } from '@care/react-icons';
import { InlineTextField, Typography } from '@care/react-component-lib';
import { AddressInput } from '@/__generated__/globalTypes';
import { logIbAddressErrorEvent } from '@/utilities/analyticsPagesUtils';

const useStyles = makeStyles(() => ({
  addressSearchContainer: {
    position: 'relative',
    marginTop: theme.spacing(-2.75),
    '&& > .MuiTextField-root': {
      paddingLeft: 0,
      paddingRight: 0,
    },
    '&& label': {
      paddingLeft: theme.spacing(5),
      textAlign: 'left',
    },
    '&& input': {
      paddingLeft: theme.spacing(7),
    },
  },
  loopIcon: {
    position: 'absolute',
    top: 42,
    left: 20,
    zIndex: 2,
  },
  locationContainer: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  locationText: {
    textAlign: 'left',
    whiteSpace: 'pre-wrap',
  },
  editLocationButton: {
    width: 81,
  },
  notFoundContainer: {
    padding: theme.spacing(5, 0),
    textAlign: 'center',
  },
  suiteContainer: {
    '&& > .MuiTextField-root': {
      paddingLeft: 0,
      paddingRight: 0,
    },
  },
}));

interface AddressSearchProps {
  seekerAddress: AddressInput;
  setSeekerAddress: (address: AddressInput) => void;
  displayAddress: string;
  setDisplayAddress: (displayAddress: string) => void;
  pageName: string;
}

const AddressSearch = ({
  seekerAddress,
  setSeekerAddress,
  displayAddress,
  setDisplayAddress,
  pageName,
}: AddressSearchProps): JSX.Element => {
  const classes = useStyles();
  const [{ placeDetails, predictions }, dispatch] = useGoogleMaps();
  const addressLabelText = 'Street address';
  const addressLabelTextOnInput = 'Search address';
  const suiteLabelText = 'Apt or Suite (optional)';
  const invalidAddressText = 'Invalid address';
  const notFoundText = 'Address not found';
  const notFoundDescription = `Make sure it's spelled correctly or try adding a city, state or zip code`;
  const CITY_POSITION: number = 2;
  const STATE_POSITION: number = 3;

  const [label, setLabel] = useState(addressLabelText);
  const [showAddressSearch, setShowAddressSearch] = useState<boolean>(true);
  const [shouldSaveAddress, setShouldSaveAddress] = useState<boolean>(false);

  const isValidAddress = seekerAddress.zip && seekerAddress.addressLine1;

  const handleFocus = () => {
    setLabel(addressLabelTextOnInput);
  };

  const handleBlur = () => {
    setLabel(addressLabelText);
  };

  const setGoogleMapsPlaceInfo = (inputText: string) => {
    dispatch({ type: GoogleMapsActions.SET_INPUT_SEARCH, payload: inputText });
    dispatch({ type: GoogleMapsActions.SET_PLACE_DETAILS, payload: null });
    dispatch({ type: GoogleMapsActions.CLEAR_PREDICTIONS });
  };

  const handleEditClick = () => {
    setGoogleMapsPlaceInfo(seekerAddress.addressLine1);
    setShowAddressSearch(true);
    setShouldSaveAddress(false);
  };

  const handlePredictionClick = () => {
    setShowAddressSearch(false);
    setShouldSaveAddress(true);
  };

  useEffect(() => {
    if (shouldSaveAddress && placeDetails) {
      setSeekerAddress({
        ...seekerAddress,
        addressLine1: placeDetails.prediction.structured_formatting.main_text,
        city: placeDetails.prediction.terms[CITY_POSITION].value,
        state: placeDetails.prediction.terms[STATE_POSITION].value,
        zip: placeDetails.zip || '',
        longitude: placeDetails.longitude,
        latitude: placeDetails.latitude,
      });
      setDisplayAddress(
        `${placeDetails.prediction.structured_formatting.main_text} \n${
          placeDetails.prediction.structured_formatting.secondary_text
        } ${placeDetails.zip || ''}`
      );
      setShouldSaveAddress(false);
    }
  }, [shouldSaveAddress, placeDetails]);

  useEffect(() => {
    if (displayAddress) {
      setShowAddressSearch(false);
    }
  }, []);

  const handleSuiteChange = (value: string) => {
    setSeekerAddress({
      ...seekerAddress,
      addressLine2: value,
    });
  };

  const addressSearchMode = showAddressSearch || shouldSaveAddress;

  useEffect(() => {
    if (addressSearchMode && !predictions) {
      logIbAddressErrorEvent(notFoundText, pageName);
    }
  }, [addressSearchMode, predictions]);
  useEffect(() => {
    if (!addressSearchMode && !isValidAddress) {
      logIbAddressErrorEvent(invalidAddressText, pageName);
    }
  }, [addressSearchMode, isValidAddress]);

  return (
    <Box>
      {addressSearchMode ? (
        <Box className={classes.addressSearchContainer}>
          <Icon24UtilitySearch className={classes.loopIcon} />
          <GoogleMaps.InputSearch label={label} onFocus={handleFocus} onBlur={handleBlur} />
          {predictions ? (
            predictions.map((prediction) => (
              <GoogleMaps.Prediction
                {...prediction}
                key={prediction.place_id}
                onClick={handlePredictionClick}
              />
            ))
          ) : (
            <Box className={classes.notFoundContainer}>
              <Typography variant="h2">{notFoundText}</Typography>
              <Typography variant="body1">{notFoundDescription}</Typography>
            </Box>
          )}
        </Box>
      ) : (
        <Box>
          <Box className={classes.locationContainer}>
            <Box className={classes.locationText}>
              <Typography>{displayAddress}</Typography>
            </Box>
            <Box>
              <Button
                color="secondary"
                variant="outlined"
                size="small"
                className={classes.editLocationButton}
                onClick={handleEditClick}>
                Edit
              </Button>
            </Box>
          </Box>
          {!isValidAddress && (
            <Box mt={2}>
              <Typography variant="body2" color="error">
                {invalidAddressText}
              </Typography>
            </Box>
          )}
          <Box className={classes.suiteContainer}>
            <InlineTextField
              id="aptOrSuite"
              InputLabelProps={{ shrink: false }}
              name="aptOrSuite"
              label={suiteLabelText}
              value={seekerAddress.addressLine2}
              onChange={(e) => handleSuiteChange(e.target.value)}
            />
          </Box>
        </Box>
      )}
    </Box>
  );
};

const AddressSearchWrapper = (props: AddressSearchProps) => {
  return (
    <GoogleMaps.Provider>
      <AddressSearch {...props} />
    </GoogleMaps.Provider>
  );
};

export default AddressSearchWrapper;
