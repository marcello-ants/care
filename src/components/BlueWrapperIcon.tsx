import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import clsx from 'clsx';

type BlueWrapperIconProps = {
  icon: React.ReactElement;
  className?: string;
  width?: number;
  height?: number;
  iconColor?: string;
  iconSize?: string;
};

const useStyles = ({ width, height }: { width: number; height: number }) => {
  return makeStyles((theme) => ({
    iconContainer: {
      backgroundColor: theme?.palette?.care?.navy[800],
      borderRadius: '50%',
      width,
      height,
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
      boxSizing: 'border-box',
      marginBottom: theme.spacing(2),
      [theme.breakpoints.up('md')]: {
        marginBottom: theme.spacing(3),
      },
    },
  }))();
};

const BlueWrapperIcon = ({
  className,
  icon,
  width = 64,
  height = 64,
  iconColor,
  iconSize,
}: BlueWrapperIconProps) => {
  const classes = useStyles({ width, height });

  return (
    <div className={clsx(classes.iconContainer, className)} data-testid="blue-icon-wrapper">
      {React.cloneElement(icon, { color: iconColor, size: iconSize })}
    </div>
  );
};

BlueWrapperIcon.defaultProps = {
  className: '',
  width: 64,
  height: 64,
  iconColor: '#fff',
  iconSize: '39px',
};

export default BlueWrapperIcon;
