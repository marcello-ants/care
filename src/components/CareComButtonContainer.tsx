/* istanbul ignore next */
import React from 'react';
import { makeStyles } from '@material-ui/core/styles';

type styleProps = {
  mt: number;
  mb: number;
};

const useStyles = makeStyles((theme) => ({
  buttonContainer: {
    padding: theme.spacing(0, 2),
    marginTop: ({ mt }: styleProps) => theme.spacing(mt),
    marginBottom: ({ mb }: styleProps) => theme.spacing(mb),
    [theme.breakpoints.up('sm')]: {
      padding: theme.spacing(0, 3),
    },
  },
}));

function CareComButtonContainer({
  children,
  mt = 0,
  mb = 0,
}: {
  children: React.ReactNode;
  mt?: number;
  mb?: number;
}) {
  const classes = useStyles({ mt, mb });

  return <div className={classes.buttonContainer}>{children}</div>;
}

CareComButtonContainer.defaultProps = {
  mt: 0,
  mb: 0,
};

export default CareComButtonContainer;
