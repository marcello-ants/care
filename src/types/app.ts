import { NextComponentType } from 'next';
import { FC, ReactElement } from 'react';
import { ClassNameMap } from '@material-ui/core/styles/withStyles';
import { SeekerAction, SeekerState } from './seeker';
import { SeekerCCAction, SeekerCCState } from './seekerCC';
import { SeekerHKAction, SeekerHKState } from './seekerHK';
import { SeekerPCAction, SeekerPCState } from './seekerPC';
import { SeekerTUAction, SeekerTUState } from './seekerTU';
import { ProviderAction, ProviderState } from './provider';
import { ProviderCCAction, ProviderCCState } from './providerCC';
import { FlowAction, FlowState } from './flow';
import { LtcgAction, LtcgState } from './ltcg';
import { JobAction } from './job';

import { ILayoutProps } from './layout';

export interface AppState {
  flow: FlowState;
  seeker: SeekerState;
  seekerCC: SeekerCCState;
  seekerHK: SeekerHKState;
  seekerPC: SeekerPCState;
  seekerTU: SeekerTUState;
  provider: ProviderState;
  providerCC: ProviderCCState;
  ltcg: LtcgState;
  version: string;
}

export type AppAction =
  | SeekerAction
  | SeekerCCAction
  | ProviderAction
  | FlowAction
  | ProviderCCAction
  | SeekerHKAction
  | JobAction
  | SeekerPCAction
  | SeekerTUAction
  | LtcgAction;

export interface AppDispatch {
  (action: AppAction): void;
}

export type AppPageComponent = NextComponentType & {
  // whether the page requires an auth check (defaults to false)
  CheckAuthCookie?: boolean;
  // custom page footer
  Footer?: ReactElement;
  // custom page layout (otherwise uses DefaultLayout)
  Layout?: FC<ILayoutProps>;
  // custom page header
  Header?: ReactElement;
  // tags that should be added to the page's <head>
  HeadTags?: ReactElement;
  // whether the page should use the default page transition (defaults to true)
  usePageTransition?: boolean;
  // when not using the default page transition, the key to use to know when to reenable the page transition (defaults to the route name)
  transitionKey?: string;
  // custom background image
  backgroundImage?: (props?: any) => ClassNameMap<'background'>;
  // whether the automatic "Screen Viewed" amplitude event logging should be disabled
  disableScreenViewed?: boolean;
};
