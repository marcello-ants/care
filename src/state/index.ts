/* eslint no-param-reassign: ["error", { "props": true, "ignorePropertyModificationsFor": ["draft"] }] */
import produce, { enableES5 } from 'immer';
import { CURRENT_STATE_VERSION, CURRENT_STATE_MAJOR_VERSION } from '../constants';
import { AppAction, AppState } from '../types/app';

import {
  SeekerAction,
  initialState as seekerInitialState,
  reducer as seekerReducer,
  validate as validateSeeker,
  doesSeekerStateHaveDefaultKeys,
} from './seeker';

import {
  SeekerCCAction,
  initialState as seekerCCInitialState,
  reducer as seekerCCReducer,
  validate as validateSeekerCC,
  doesSeekerCCStateHaveDefaultKeys,
} from './seekerCC';

import {
  ProviderAction,
  initialState as providerInitialState,
  reducer as providerReducer,
} from './provider';

import {
  ProviderCCAction,
  initialState as providerCCInitialState,
  reducer as providerCCReducer,
} from './providerCC';

import {
  SeekerHKAction,
  initialState as seekerHKInitialState,
  reducer as seekerHKReducer,
} from './seekerHK';

import {
  SeekerPCAction,
  initialState as seekerPCInitialState,
  reducer as seekerPCReducer,
} from './seekerPC';

import {
  SeekerTUAction,
  initialState as seekerTUInitialState,
  reducer as seekerTUReducer,
} from './seekerTU';

import { FlowAction, initialState as flowInitialState, reducer as flowReducer } from './flow';

import { LtcgAction, initialState as ltcgInitialState, reducer as ltcgReducer } from './ltcg';

// immer uses Proxy, but if that's not available the ES5 plugin needs to be enabled
/* istanbul ignore if */
if (typeof Proxy === 'undefined') {
  // thanks, IE 11
  enableES5();
}

export const initialAppState: AppState = {
  flow: flowInitialState,
  seeker: seekerInitialState,
  seekerCC: seekerCCInitialState,
  provider: providerInitialState,
  providerCC: providerCCInitialState,
  seekerHK: seekerHKInitialState,
  seekerPC: seekerPCInitialState,
  seekerTU: seekerTUInitialState,
  ltcg: ltcgInitialState,
  version: CURRENT_STATE_VERSION,
};

export const rootReducer = produce((draft: AppState, action: AppAction) => {
  draft.seeker = seekerReducer(draft.seeker, action as SeekerAction);
  draft.provider = providerReducer(draft.provider, action as ProviderAction);
  draft.providerCC = providerCCReducer(draft.providerCC, action as ProviderCCAction);
  draft.flow = flowReducer(draft.flow, action as FlowAction);
  draft.seekerCC = seekerCCReducer(draft.seekerCC, action as SeekerCCAction);
  draft.seekerHK = seekerHKReducer(draft.seekerHK, action as SeekerHKAction);
  draft.seekerPC = seekerPCReducer(draft.seekerPC, action as SeekerPCAction);
  draft.seekerTU = seekerTUReducer(draft.seekerTU, action as SeekerTUAction);
  draft.ltcg = ltcgReducer(draft.ltcg, action as LtcgAction);

  return draft;
});

export function validateState(state: AppState, pathName: string) {
  // TODO: call jobPost validation here instead of in AppState
  return validateSeeker(state, pathName) && validateSeekerCC(state, pathName);
}

export function isStateVersionCompatible(state: AppState) {
  if (state?.version) {
    const versionComponents = state.version.split('.');
    if (versionComponents.length === 3 && versionComponents[0] === CURRENT_STATE_MAJOR_VERSION) {
      return true;
    }
  }
  return false;
}

export function doesStateHaveDefaultKeys(state: AppState) {
  return (
    doesSeekerStateHaveDefaultKeys(state.seeker) && doesSeekerCCStateHaveDefaultKeys(state.seekerCC)
  );
}
