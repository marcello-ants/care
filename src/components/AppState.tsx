import { useRouter } from 'next/router';
import { createContext, useReducer, useContext, ReactNode, useEffect } from 'react';
import { debounce } from 'lodash-es';
import { configureScope, captureException, Severity } from '@sentry/nextjs';
import {
  initialAppState,
  rootReducer,
  isStateVersionCompatible,
  doesStateHaveDefaultKeys,
  validateState,
} from '@/state';
import patchAppState from '@/state/patch';
import { initialState as flowInitialState } from '@/state/flow';
import logger from '@/lib/clientLogger';
import { ClientLogger } from '@care/utils';
import getVertical from '@/utilities/verticalUtils';
import { SessionStorageHelper } from '@/utilities/SessionStorageHelper';
import { trackPageViewInSiftScience } from '@/utilities/siftScienceHelper';
import { AppAction, AppDispatch, AppState } from '@/types/app';
import { StateCheckingResponse } from '@/types/common';
import {
  LOCAL_STORAGE_STATE_KEY,
  CZEN_BASE_PATH,
  FLOWS,
  POST_A_JOB_ROUTES,
  CLIENT_FEATURE_FLAGS,
} from '@/constants';
import { useFeatureFlags } from '@/components/FeatureFlagsContext';
import { AnalyticsHelper } from '@/utilities/analyticsHelper';

type AppStateProviderProps = {
  children: ReactNode;
  initialStateOverride?: AppState;
  czenJSessionId?: string | undefined;
  referrerCookie?: string | undefined | null;
  userHasAccount?: boolean;
  memberUuid?: string;
};

function getStateFromStorageOrDefault(defaultState: AppState) {
  let stateFromStorageJson;
  const sessionStorageAvailable = SessionStorageHelper.isAvailable();

  if (sessionStorageAvailable) {
    stateFromStorageJson = SessionStorageHelper.getItem(LOCAL_STORAGE_STATE_KEY);
  } else {
    logger.warn({ event: 'sessionStorageUnavailable' });
  }

  if (stateFromStorageJson) {
    const stateFromStorage = JSON.parse(stateFromStorageJson) as AppState;
    try {
      const patchedState = patchAppState(stateFromStorage, defaultState);

      if (isStateVersionCompatible(patchedState) && doesStateHaveDefaultKeys(patchedState)) {
        return patchedState;
      }
      logger.warn({ event: 'invalidStateFromLocalStorage', state: stateFromStorageJson });
    } catch (e) {
      captureException(e, { level: Severity.Warning });
    }
  }
  return defaultState;
}

function persistState(state: AppState) {
  SessionStorageHelper.setItem(LOCAL_STORAGE_STATE_KEY, JSON.stringify(state));
}

const debouncePersistState = debounce(persistState, 500, {
  leading: true,
  trailing: true,
  maxWait: 1000,
});

function reducerWrapper(state: AppState, action: AppAction) {
  const nextState = rootReducer(state, action);
  debouncePersistState(nextState);
  return nextState;
}

export function redirectIfInvalidState(state: AppState, pathname: string): StateCheckingResponse {
  if (
    typeof Intl === 'undefined' ||
    typeof Intl.NumberFormat === 'undefined' ||
    typeof Intl.DateTimeFormat === 'undefined'
  ) {
    // TODO: Remove this with SC-214
    logger.error({ event: 'IntlUnavailable', source: pathname, invalidState: state });
    return {
      redirectCalled: true,
      readableMessage: 'Safari 9 and lower is not currently supported',
    };
  }

  const isStateValid = validateState(state, pathname);
  const redirect = !isStateValid;

  if (redirect) {
    SessionStorageHelper.removeItem(LOCAL_STORAGE_STATE_KEY);

    window.location.assign(CZEN_BASE_PATH);
  }

  return { redirectCalled: redirect };
}

export function getFlowName(pathname: string, oldFlowName?: string): string {
  let newFlowName = Object.values(FLOWS).find((flow) => pathname.includes(flow.path))?.name;

  if (!newFlowName) {
    // TODO: this shouldn't be needed after SC-123 is implemented
    const isPostAJob = Object.values(POST_A_JOB_ROUTES).some((route) => pathname.includes(route));
    if (isPostAJob) {
      newFlowName = 'SEEKER_POST_A_JOB';
    }
  }

  // Don't allow shifting from in-facility to seeker flow name
  if (newFlowName === FLOWS.SEEKER.name && oldFlowName === FLOWS.SEEKER_IN_FACILITY.name) {
    return FLOWS.SEEKER_IN_FACILITY.name;
  }

  return newFlowName ?? '';
}

function useFlowDetection(dispatch: AppDispatch, oldFlowName: string) {
  const { pathname } = useRouter();
  const newFlowName = getFlowName(pathname, oldFlowName);
  const vertical = getVertical();
  ClientLogger.setLoggerStatics({ flow: newFlowName, vertical });

  useEffect(() => {
    dispatch({ type: 'setFlowName', flowName: newFlowName });

    configureScope((scope) => {
      scope.setTags({
        flow: newFlowName,
        vertical,
      });
    });
  }, [newFlowName]);
}

const AppStateContext = createContext<AppState>(initialAppState);
const AppDispatchContext = createContext<AppDispatch>(() =>
  // eslint-disable-next-line no-console
  console.warn(
    'Calling default dispatch is a no-op! You are probably consuming app dispatch in a component above AppStateProvider'
  )
);

export const AppStateProvider = ({
  children,
  initialStateOverride,
  czenJSessionId,
  referrerCookie,
  userHasAccount,
  memberUuid,
}: AppStateProviderProps) => {
  const hydratedInitialAppState = {
    ...initialAppState,
    flow: {
      ...flowInitialState,
      czenJSessionId,
      referrerCookie,
      userHasAccount,
      memberUuid,
    },
  };
  const featureFlags = useFeatureFlags();
  const sessionStorageRetrievalSuccessEvaluation =
    featureFlags?.flags[CLIENT_FEATURE_FLAGS.SESSION_STORAGE_RETRIEVAL_SUCCESS];
  const shouldMeasureSessionStorageRetrievalSuccess =
    sessionStorageRetrievalSuccessEvaluation?.variationIndex === 1;

  const router = useRouter();

  const logSessionStorageError = () => {
    logger.info({ event: 'sessionStorageStateUnavailable', path: router.pathname });
    AnalyticsHelper.logEvent({
      name: 'sessionStorageStateUnavailable',
    });
  };
  const launchDarklyWrappedGetStateFromStorageOrDefault = (defaultState: AppState) => {
    if (shouldMeasureSessionStorageRetrievalSuccess) {
      try {
        const sessionStorageState = SessionStorageHelper.getItem(LOCAL_STORAGE_STATE_KEY);
        if (sessionStorageState && sessionStorageState.length > 0) {
          const stateFromStorage = JSON.parse(sessionStorageState);
          if (stateFromStorage) {
            logger.info({ event: 'sessionStorageStateAvailable', path: router.pathname });
          } else {
            logSessionStorageError();
          }
        } else {
          logSessionStorageError();
        }
      } catch (_e) {
        logSessionStorageError();
      }
    }
    return getStateFromStorageOrDefault(defaultState);
  };

  const [state, dispatch] = useReducer(
    reducerWrapper,
    initialStateOverride ?? hydratedInitialAppState,
    launchDarklyWrappedGetStateFromStorageOrDefault
  );
  useFlowDetection(dispatch, state.flow.flowName);
  const routePath = router.asPath?.split('?')[0];
  const stateCheckResult = redirectIfInvalidState(state, routePath);
  if (stateCheckResult.redirectCalled) {
    if (stateCheckResult.readableMessage) {
      return (
        <>
          <p>Redirecting to Member Home Page...</p>
          <p>{stateCheckResult.readableMessage}</p>
        </>
      );
    }
    return <></>;
  }

  useEffect(() => {
    trackPageViewInSiftScience(state);
  }, [routePath]);

  return (
    <AppDispatchContext.Provider value={dispatch}>
      <AppStateContext.Provider value={state}>{children}</AppStateContext.Provider>
    </AppDispatchContext.Provider>
  );
};

AppStateProvider.defaultProps = {
  initialStateOverride: undefined,
  czenJSessionId: undefined,
  referrerCookie: undefined,
  userHasAccount: false,
  memberUuid: undefined,
};

export const useAppState = () => useContext(AppStateContext);
export const useAppDispatch = () => useContext(AppDispatchContext);

export const useSeekerState = () => useContext(AppStateContext).seeker;
export const useProviderState = () => useContext(AppStateContext).provider;
export const useProviderCCState = () => useContext(AppStateContext).providerCC;
export const useFlowState = () => useContext(AppStateContext).flow;
export const useSeekerCCState = () => useContext(AppStateContext).seekerCC;
export const useSeekerHKState = () => useContext(AppStateContext).seekerHK;
export const useSeekerPCState = () => useContext(AppStateContext).seekerPC;
export const useSeekerTUState = () => useContext(AppStateContext).seekerTU;
export const useLtcgState = () => useContext(AppStateContext).ltcg;
