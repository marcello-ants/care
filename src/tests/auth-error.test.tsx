import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import { useRouter } from 'next/router';
import AuthError from '@/pages/auth-error';
import { AppStateProvider } from '@/components/AppState';
import { initialAppState } from '@/state';
import { AppState } from '@/types/app';
import logger from '@/lib/clientLogger';

jest.mock('next/router', () => ({
  __esModule: true,
  useRouter: jest.fn(),
}));
const windowHistoryBackMock = jest.fn();
const loggerInfoMock = jest.fn();

describe('auth-error page', () => {
  let mockRouter: any = null;
  let asFragment: any | null = null;

  const renderComponent = (initialState: AppState) => {
    const view = render(
      <AppStateProvider initialStateOverride={initialState}>
        <AuthError />
      </AppStateProvider>
    );
    ({ asFragment } = view);
  };

  beforeAll(() => {
    // @ts-ignore
    delete window.history;
    // @ts-ignore
    window.history = {
      back: windowHistoryBackMock,
    };
    logger.info = loggerInfoMock;
  });

  beforeEach(() => {
    mockRouter = {
      push: jest.fn(), // the component uses `router.push` only
      pathname: '/auth-error',
    };
    (useRouter as jest.Mock).mockReturnValue(mockRouter);
  });

  afterEach(() => {
    // cleanup on exiting
    mockRouter = null;
    asFragment = null;
    loggerInfoMock.mockReset();
  });

  it('matches snapshot', () => {
    renderComponent(initialAppState);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly', () => {
    renderComponent(initialAppState);
    expect(screen.getByText(/It looks like you're already logged in./i)).toBeInTheDocument();
  });

  it('logout and continue', () => {
    renderComponent(initialAppState);
    const logoutButton = screen.getByText(/Logout and continue/i);
    fireEvent.click(logoutButton);
    expect(loggerInfoMock).toHaveBeenCalledWith({
      event: 'logOutClick',
      page: 'authError',
    });
  });

  it('should go to back', () => {
    renderComponent(initialAppState);
    const goBackButton = screen.getByText(/Go back/i);
    fireEvent.click(goBackButton);
    expect(windowHistoryBackMock).toHaveBeenCalled();
  });
});
