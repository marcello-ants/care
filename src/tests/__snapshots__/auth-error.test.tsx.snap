// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth-error page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root makeStyles-root-1 MuiGrid-container"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root makeStyles-root-7 makeStyles-icon-2"
        focusable="false"
        viewBox="0 0 64 64"
      >
        <path
          d="M41.41 42.82a1.4 1.4 0 01-1-.41L21.59 23.58a1.41 1.41 0 012-2l18.82 18.84a1.41 1.41 0 010 2 1.44 1.44 0 01-1 .4z"
          fill="#213f5f"
        />
        <path
          d="M22.59 42.82a1.44 1.44 0 01-1-.41 1.41 1.41 0 010-2l18.83-18.82a1.41 1.41 0 012 2L23.58 42.41a1.4 1.4 0 01-.99.41z"
          fill="#213f5f"
        />
        <path
          d="M32 58.7A26.7 26.7 0 1158.7 32 26.73 26.73 0 0132 58.7zm0-50.58A23.88 23.88 0 1055.88 32 23.9 23.9 0 0032 8.12z"
          fill="#213f5f"
        />
      </svg>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <h2
        class="MuiTypography-root makeStyles-header-8 MuiTypography-h2"
      >
        It looks like you're already logged in.
      </h2>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <p
        class="MuiTypography-root makeStyles-bodyText-3 MuiTypography-body2"
      >
        In order to start, you'll need to logout and create an account with a separate email address.
      </p>
    </div>
    <div
      class="MuiGrid-root makeStyles-buttonContainer-4 MuiGrid-item MuiGrid-grid-xs-12"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge MuiButton-fullWidth"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <span>
            Logout and continue
          </span>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-skipButton-5 MuiButton-textSizeLarge MuiButton-sizeLarge MuiButton-fullWidth"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <span>
            Go back
          </span>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
